enhancements/cert-manager/istio-csr-controller.md

```diff
@@ -0,0 +1,929 @@
+	// +kubebuilder:example:=["cert-manager-istio-csr.istio-system.svc"]
+	// +optional
+	CertificateDNSNames []string `json:"certificateDNSNames,omitempty"`
+
+	// IstiodCertificateDuration is the istio's certificate validity period.
+	// +kubebuilder:default:="1h"
+	// +optional
+	IstiodCertificateDuration time.Duration `json:"istiodCertificateDuration,omitempty"`
+
+	// IstiodCertificateRenewBefore is the ahead time to renew the istio's certificate before
+	// expiry.
+	// +kubebuilder:default:="30m"
+	// +optional
+	IstiodCertificateRenewBefore time.Duration `json:"istiodCertificateRenewBefore,omitempty"`
+
+	// IstiodPrivateKeySize is the key size to be for RSAKey.
+	// +kubebuilder:default:=2048
+	// +optional
+	IstiodPrivateKeySize int32 `json:"istiodPrivateKeySize,omitempty"`
+}
+
+// ServerConfig is for configuring the server endpoint used by istio
+// for ontaining the certificates.
+type ServerConfig struct {
+	// Address to serve istio-csr gRPC service.
+	// +kubebuilder:default:="0.0.0.0"
+	// +optional
+	Address string `json:"address,omitempty"`
+
+	// Port to serve istio-csr gRPC service.
+	// +kubebuilder:default:="6443"
+	// +optional
+	Port string `json:"port,omitempty"`
+
+	// CertificateKeySize is the server's serving certificate's key size.
+	// +kubebuilder:default:=2048
+	// +optional
+	CertificateKeySize int32 `json:"certificateKeySize,omitempty"`
+
+	// SignatureAlgorithm is the signature algorithm to use when generating
+	// private keys. At present only RSA and ECDSA are supported.
+	// +kubebuilder:default:="RSA"
+	// +kubebuilder:validation:Enum:="RSA";"ECDSA"
+	// +optional
+	SignatureAlgorithm string `json:"signatureAlgorithm,omitempty"`
+
+	// MaxCertificateDuration is the maximum validity duration that can be
+	// requested for a certificate.
+	// +kubebuilder:default:="1h"
+	// +optional
+	MaxCertificateDuration time.Duration `json:"certificateDuration,omitempty"`
+}
+
+// IstioConfig is for configuring the istio specifics.
+type IstioConfig struct {
+	// Revisions are the istio revisions that are currently installed in the cluster.
+	// Changing this field will modify the DNS names that will be requested for
+	// the istiod certificate.
+	// +kubebuilder:default:=["basic"]
+	// +optional
+	Revisions []string `json:"revisions,omitempty"`
+
+	// namespace of the istio control-plane. In the same namespace issuer will be created
+	// used for obtaining the serving certificates.
+	Namespace string `json:"namespace,omitempty"`
+}
+
+// ControllerConfig is for configuring the controller for setting up
+// defaults to enable istio-csr agent.
+type ControllerConfig struct {
+	// labels to apply to all resources created for istio-csr agent deployment.
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// issuerConfig is for setting the issuer specifics to be created by the controller.
+	IssuerConfig *IssuerConfig `json:"issuerConfig,omitempty"`
```

Comment:
Is this issuer created automatically by istio-csr as default or operator? It'd be best to let user add the CR rather than replicating the same here, if possible.

Reply:
suggestion incorporated.